File: packages/hermes-gateway/src/paperclip-client.ts
import type { GatewayConfig } from "./config.js";
import type { PaperclipIssue } from "./types.js";

export class PaperclipClient {
  constructor(private readonly config: GatewayConfig) {}

  private get headers(): Record<string, string> {
    return {
      "Content-Type": "application/json",
      Authorization: `Bearer ${this.config.paperclipApiKey}`,
    };
  }

  private url(path: string): string {
    return `${this.config.paperclipApiUrl}${path}`;
  }

  async createIssue(params: {
    title: string;
    description: string;
    priority?: string;
    metadata?: Record<string, unknown>;
  }): Promise<PaperclipIssue> {
    const res = await fetch(
      this.url(`/api/companies/${this.config.paperclipCompanyId}/issues`),
      {
        method: "POST",
        headers: this.headers,
        body: JSON.stringify({
          title: params.title,
          description: params.description,
          priority: params.priority || "medium",
          metadata: params.metadata,
        }),
      },
    );

    if (!res.ok) {
      const text = await res.text();
      throw new Error(`Paperclip createIssue failed (${res.status}): ${text}`);
    }

    return (await res.json()) as PaperclipIssue;
  }

  async addComment(issueId: string, body: string): Promise<void> {
    const res = await fetch(this.url(`/api/issues/${issueId}/comments`), {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify({ body }),
    });

    if (!res.ok) {
      const text = await res.text();
      throw new Error(`Paperclip addComment failed (${res.status}): ${text}`);
    }
  }

  async getIssue(issueId: string): Promise<PaperclipIssue> {
    const res = await fetch(this.url(`/api/issues/${issueId}`), {
      method: "GET",
      headers: this.headers,
    });

    if (!res.ok) {
      const text = await res.text();
      throw new Error(`Paperclip getIssue failed (${res.status}): ${text}`);
    }

    return (await res.json()) as PaperclipIssue;
  }

  async registerWebhook(params: {
    url: string;
    events: string[];
    secret: string;
  }): Promise<{ id: string }> {
    const res = await fetch(
      this.url(`/api/companies/${this.config.paperclipCompanyId}/webhooks`),
      {
        method: "POST",
        headers: this.headers,
        body: JSON.stringify(params),
      },
    );

    if (!res.ok) {
      const text = await res.text();
      throw new Error(`Paperclip registerWebhook failed (${res.status}): ${text}`);
    }

    return (await res.json()) as { id: string };
  }
}

File: tools/n8n-workflow-watcher/paperclip_client.py
"""Dünner, stdlib-only Client für die Paperclip Control-Plane (:3100)."""
from __future__ import annotations

import json
import os
import urllib.error
import urllib.request

DEFAULT_BASE = "http://localhost:3100"


class ApiError(RuntimeError):
    pass


def load_token(auth_path: str | None = None) -> str:
    auth_path = auth_path or os.path.expanduser("~/.paperclip/auth.json")
    try:
        with open(auth_path) as fh:
            data = json.load(fh)
    except (OSError, ValueError):
        return ""
    creds = (data or {}).get("credentials", {})
    entry = creds.get(DEFAULT_BASE) or creds.get("http://127.0.0.1:3100") or {}
    return entry.get("token", "")


def _post(url: str, token: str, payload: dict) -> dict:
    req = urllib.request.Request(
        url,
        data=json.dumps(payload).encode("utf-8"),
        headers={"Content-Type": "application/json",
                 "Authorization": f"Bearer {token}"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=30) as r:
            raw = r.read().decode("utf-8") or "{}"
            return json.loads(raw)
    except urllib.error.HTTPError as e:
        raise ApiError(f"HTTP {e.code} for {url}") from e
    except Exception as e:  # noqa: BLE001
        raise ApiError(f"request failed for {url}: {e}") from e


def create_issue(base: str, token: str, company_id: str, *, title: str,
                 description: str, assignee_agent_id: str | None,
                 priority: str = "medium", parent_id: str | None = None) -> str:
    payload = {"title": title, "description": description, "priority": priority}
    if assignee_agent_id:
        payload["assigneeAgentId"] = assignee_agent_id
    if parent_id:
        payload["parentId"] = parent_id
    out = _post(f"{base}/api/companies/{company_id}/issues", token, payload)
    return out.get("id", "")


def add_comment(base: str, token: str, issue_id: str, body: str) -> None:
    _post(f"{base}/api/issues/{issue_id}/comments", token, {"body": body})
